diff -ur gnutls-3.3.8b1.orig/lib/fips.c gnutls-3.3.8b1/lib/fips.c
--- gnutls-3.3.8b1.orig/lib/fips.c	2014-09-03 16:53:13.000000000 +0200
+++ gnutls-3.3.8b1/lib/fips.c	2014-09-04 17:11:26.825733730 +0200
@@ -97,7 +97,7 @@
 #define HOGWEED_LIBRARY_NAME "libhogweed.so.2"
 #define GMP_LIBRARY_NAME "libgmp.so.10"
 
-static const char fips_key[] = "I'd rather be skiing";
+static const char fips_key[] = "orboDeJITITejsirpADONivirpUkvarP";
 
 #define HMAC_SUFFIX ".hmac"
 #define HMAC_SIZE 32
