diff --git a/tests/Makefile.am b/tests/Makefile.am
index d249d405f..6dc63758d 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -156,7 +156,11 @@ endif
 endif
 
 check_PROGRAMS = $(ctests)
-dist_check_SCRIPTS = rfc2253-escape-test sni-hostname.sh
+dist_check_SCRIPTS = rfc2253-escape-test
+
+if !WINDOWS
+dist_check_SCRIPTS += sni-hostname.sh
+endif
 
 TESTS = $(ctests) $(dist_check_SCRIPTS)
 
-- 
2.14.3

