diff --git a/lib/accelerated/x86/x86-common.c b/lib/accelerated/x86/x86-common.c
index cc67b08..e730ba0 100644
--- a/lib/accelerated/x86/x86-common.c
+++ b/lib/accelerated/x86/x86-common.c
@@ -37,6 +37,7 @@
 # include <sha-padlock.h>
 #endif
 #include <aes-padlock.h>
+#include <fips.h>
 
 /* ebx, ecx, edx 
  * This is a format compatible with openssl's CPUID detection.
@@ -581,7 +582,8 @@ void register_x86_crypto(void)
 	
 	register_x86_intel_crypto(capabilities);
 #ifdef ENABLE_PADLOCK
-	register_x86_padlock_crypto(capabilities);
+	if (_gnutls_fips_mode_enabled() != 1)
+		register_x86_padlock_crypto(capabilities);
 #endif
 }
 
